File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/card/application/adapter/PaymentAdapter.java
package com.example.accountservice.card.application.adapter;

import java.time.LocalDateTime;

/**
 * Adapter interface (ACL) for synchronously querying the Payment BC. The monthly card usage
 * statement notification (scheduling.md Feature 2) needs to look up payment history from the
 * Payment BC to calculate each card's transaction count and total, so a synchronous Adapter pattern
 * is used (see cross-domain.md — same reasoning as card/application/adapter/AccountAdapter).
 *
 * <p>The return type is translated into the minimal aggregated shape the Card BC needs ({@link
 * UsageSummary}) — it does not expose the Payment BC's Payment entity or its status enum.
 */
public interface PaymentAdapter {

    UsageSummary summarizeUsage(String cardId, LocalDateTime from, LocalDateTime to);

    /** The minimal usage-summary view owned by the Card BC. */
    record UsageSummary(long count, long totalAmount) {}
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/card/application/adapter/PaymentAdapter.kt
package com.example.accountservice.card.application.adapter

import java.time.LocalDateTime

/**
 * Adapter interface (Anticorruption Layer) for synchronously querying the Payment BC.
 *
 * Used by the monthly card-usage-statement (SendMonthlyCardStatementsService) to aggregate the
 * payment count/total per card — this reuses the same pattern the Card BC already uses to query
 * Account ([com.example.accountservice.card.application.adapter.AccountAdapter]). The return type
 * does not expose the Payment BC's `Payment`/`PaymentStatus`; it is translated into the minimal form
 * the Card BC needs ([PaymentSummaryView]). The actual translation is handled by
 * [com.example.accountservice.card.infrastructure.CardPaymentAdapterImpl].
 */
interface PaymentAdapter {
    fun summarizePayments(
        cardId: String,
        from: LocalDateTime,
        to: LocalDateTime,
    ): PaymentSummaryView
}

data class PaymentSummaryView(
    val count: Long,
    val totalAmount: Long,
)
